File: atlas_extracter.py
from __future__ import annotations
import logging
from dataclasses import dataclass, field
from io import BytesIO
from pathlib import Path
from typing import Dict, List, Mapping, Optional, Tuple, Union

from PIL import Image

logging.basicConfig(level=logging.INFO, format='[%(levelname)s] %(message)s')

@dataclass
class AtlasRegion:
    name: str           # unique key (e.g. "arm#2" for a duplicate region name)
    atlas_name: str     # real name written to the atlas (e.g. "arm")
    page_filename: str
    index: int = -1
    x: int = 0
    y: int = 0
    w: int = 0
    h: int = 0
    offsets: Optional[Tuple[int, int, int, int]] = None
    rotate: int = 0
    split: Optional[List[int]] = None
    pad: Optional[List[int]] = None
    extra_pairs: List[Tuple[str, List[str]]] = field(default_factory=list)

@dataclass
class AtlasPage:
    filename: str
    size: Tuple[int, int] = (0, 0)
    format: str = "RGBA8888"
    filter: Tuple[str, str] = ("Nearest", "Nearest")
    repeat: str = "none"
    pma: bool = False
    scale_x: float = 1.0
    scale_y: float = 1.0

class AtlasProcessor:
    def __init__(self, atlas_content: str, image_loader: Mapping[str, Union[str, bytes, Path, Image.Image]]):
        self.atlas_content = atlas_content
        self.pages: List[AtlasPage] = []
        self.regions: Dict[str, AtlasRegion] = {} 
        self._loaded_images: Dict[str, Image.Image] = {} 
        self._page_map: Dict[str, AtlasPage] = {}
        self._cache: Dict[str, Image.Image] = {} # Cache for extracted regions
        
        self._parse_atlas()
        if image_loader:
            self._load_images(image_loader)

    def _parse_atlas(self) -> None:
        lines = [line.strip() for line in self.atlas_content.splitlines()]
        iterator = iter(lines)
        
        current_page: Optional[AtlasPage] = None
        current_region: Optional[AtlasRegion] = None
        region_name_counts: Dict[str, int] = {}

        def get_unique_region_key(atlas_name: str) -> str:
            nxt = region_name_counts.get(atlas_name, 0) + 1
            region_name_counts[atlas_name] = nxt
            return atlas_name if nxt == 1 else f"{atlas_name}#{nxt}"

        while True:
            try:
                line = next(iterator)
            except StopIteration:
                break

            if not line:
                # --- จุดที่แก้ไข --- 
                # เจอบรรทัดว่าง "ห้าม" Reset current_page หรือ current_region ทิ้ง
                # เพราะบางไฟล์มีบรรทัดว่างคั่นระหว่าง Name กับ Properties
                continue

            # 1. Check Page (ends with .png)
            if line.endswith('.png'):
                current_page = AtlasPage(filename=line)
                self.pages.append(current_page)
                self._page_map[line] = current_page
                current_region = None # New page starts, reset region context
                continue

            # 2. Check Key-Value Pair
            if ':' in line:
                key, value_str = line.split(':', 1)
                key = key.strip().lower()
                values = [v.strip() for v in value_str.split(',')]

                if current_region:
                    # Region Properties
                    if key == 'bounds':
                        if len(values) >= 4:
                            current_region.x = int(values[0])
                            current_region.y = int(values[1])
                            current_region.w = int(values[2])
                            current_region.h = int(values[3])
                    elif key == 'xy': # Support LibGDX old format
                        current_region.x = int(values[0])
                        current_region.y = int(values[1])
                    elif key == 'size' and current_region.w == 0:
                        # Only apply size to region if we haven't got bounds yet
                        current_region.w = int(values[0])
                        current_region.h = int(values[1])
                    elif key == 'rotate':
                        val = values[0].lower()
                        if val == 'true':
                            current_region.rotate = 90
                        elif val == 'false':
                            current_region.rotate = 0
                        else:
                            try:
                                current_region.rotate = int(val)
                            except ValueError:
                                current_region.rotate = 0
                    elif key == 'offsets':
                        if len(values) >= 4:
                            current_region.offsets = tuple(map(int, values[:4])) # type: ignore
                    elif key == 'index':
                        current_region.index = int(values[0])
                    elif key == 'split' and len(values) >= 4:
                        current_region.split = [int(v) for v in values]
                    elif key == 'pad' and len(values) >= 4:
                        current_region.pad = [int(v) for v in values]
                    else:
                        current_region.extra_pairs.append((key, list(values)))

                elif current_page:
                    if key == 'size':
                        current_page.size = (int(values[0]), int(values[1]))
                    elif key == 'format':
                        current_page.format = values[0]
                    elif key == 'filter':
                        current_page.filter = (values[0], values[1])
                    elif key == 'repeat':
                        current_page.repeat = values[0]
                    elif key == 'pma':
                        current_page.pma = str(values[0]).strip().lower() == 'true'

            else:
                # 3. If no colon and not .png -> It's a Region Name
                if current_page is None:
                    continue
                
                # Found a new region name -> Create object.
                # Duplicate names (common with Spine index: multi-region sprites)
                # get a unique dict key; the real name is kept in atlas_name.
                region_key = get_unique_region_key(line)
                current_region = AtlasRegion(
                    name=region_key,
                    atlas_name=line,
                    page_filename=current_page.filename,
                )
                self.regions[region_key] = current_region

    def _load_images(self, loader: Mapping[str, Union[str, bytes, Path, Image.Image]]) -> None:
        for page in self.pages:
            source = loader.get(page.filename)
            if source is None:
                for key, val in loader.items():
                    if page.filename in str(key):
                        source = val
                        break
            
            if source is None:
                logging.debug(f"❌ Image NOT FOUND for page: {page.filename}, skipping...")
                continue

            try:
                if isinstance(source, (str, Path)):
                    img = Image.open(source).convert('RGBA')
                elif isinstance(source, bytes):
                    img = Image.open(BytesIO(source)).convert('RGBA')
                elif isinstance(source, Image.Image):
                    img = source.convert('RGBA')
                else:
                    logging.error(f"Unsupported image source type for {page.filename}: {type(source)}")
                    continue
                
                # Auto Scale Check (scale atlas coords to match real image)
                if page.size != (0, 0):
                    atlas_w, atlas_h = page.size
                    real_w, real_h = img.size
                    
                    if real_w != atlas_w or real_h != atlas_h:
                        page.scale_x = real_w / atlas_w
                        page.scale_y = real_h / atlas_h
                        logging.warning(
                            f"⚠️ Scale Mismatch: Atlas={atlas_w}x{atlas_h}, "
                            f"Real={real_w}x{real_h}. "
                            f"Scaling atlas coords (x{page.scale_x:.3f}, x{page.scale_y:.3f})"
                        )
                
                self._loaded_images[page.filename] = img
                logging.info(f"✅ Loaded {page.filename} ({img.size})")
                
            except Exception as e:
                logging.error(f"Failed to load image {page.filename}: {e}")

    @staticmethod
    def crop_and_rotate(
        image: Image.Image, x: int, y: int, w: int, h: int, rotate: int,
    ) -> Image.Image:
        """Crop a region from *image* and undo atlas rotation.

        Args:
            image: Source atlas image.
            x, y: Top-left position of the region in the atlas.
            w, h: Region dimensions (before atlas rotation is applied).
            rotate: Atlas rotation value (0, 90, 180, 270).

        Returns:
            The cropped sprite in its original (unrotated) orientation.
        """
        # When rotated 90/270, the atlas stores w/h swapped
        crop_w = h if rotate in (90, 270) else w
        crop_h = w if rotate in (90, 270) else h

        sprite = image.crop((x, y, x + crop_w, y + crop_h))

        if rotate == 90:
            sprite = sprite.transpose(Image.Transpose.ROTATE_270)
        elif rotate == 270:
            sprite = sprite.transpose(Image.Transpose.ROTATE_90)
        elif rotate == 180:
            sprite = sprite.transpose(Image.Transpose.ROTATE_180)

        return sprite

    def get_page_image(self, page_filename: Optional[str] = None) -> Optional[Image.Image]:
        """Get a loaded page image by filename, or the first one if not specified."""
        if page_filename:
            return self._loaded_images.get(page_filename)
        if self._loaded_images:
            return next(iter(self._loaded_images.values()))
        return None

    def extract_region(self, name: str) -> Optional[Image.Image]:
        region = self.regions.get(name)
        if not region: return None

        base_img = self._loaded_images.get(region.page_filename)
        if not base_img: return None

        x, y, raw_w, raw_h = region.x, region.y, region.w, region.h
        rot = region.rotate

        # Apply page scale factors (atlas coords → real image coords)
        page = self._page_map.get(region.page_filename)
        if page and (page.scale_x != 1.0 or page.scale_y != 1.0):
            sx, sy = page.scale_x, page.scale_y
            x = round(x * sx)
            y = round(y * sy)
            raw_w = round(raw_w * sx)
            raw_h = round(raw_h * sy)

        sprite = self.crop_and_rotate(base_img, x, y, raw_w, raw_h, rot)
        current_w, current_h = sprite.size

        # Offsets
        if region.offsets:
            off_x, off_y, orig_w, orig_h = region.offsets

            # Scale offsets to match real image
            if page and (page.scale_x != 1.0 or page.scale_y != 1.0):
                sx, sy = page.scale_x, page.scale_y
                off_x = round(off_x * sx)
                off_y = round(off_y * sy)
                orig_w = round(orig_w * sx)
                orig_h = round(orig_h * sy)

            canvas = Image.new('RGBA', (orig_w, orig_h), (0, 0, 0, 0))
            paste_x = off_x
            paste_y = orig_h - off_y - current_h

            canvas.paste(sprite, (paste_x, paste_y))
            self._cache[name] = canvas
            return canvas
        else:
            self._cache[name] = sprite
            return sprite

    def extract_all(self) -> List[Tuple[str, Image.Image]]:
        results = []
        for name in self.regions:
            try:
                img = self.extract_region(name)
                if img:
                    results.append((name, img))
            except Exception as e:
                logging.error(f"Failed to extract {name}: {e}")
        return results

File: android-app/www/js/atlas-extracter.js
/**
 * atlas-extracter.js
 * Port of atlas_extracter.py using Canvas API.
 */

class AtlasPage {
  constructor(filename) {
    this.filename = filename;
    this.width = 0;
    this.height = 0;
    this.format = 'RGBA8888';
    this.filter = ['Nearest', 'Nearest'];
    this.repeat = 'none';
    this.scaleX = 1.0;
    this.scaleY = 1.0;
  }
}

class AtlasRegion {
  constructor(name, pageFilename) {
    this.name = name;
    this.pageFilename = pageFilename;
    this.index = -1;
    this.x = 0;
    this.y = 0;
    this.w = 0;
    this.h = 0;
    this.offsets = null; // [off_x, off_y, orig_w, orig_h]
    this.rotate = 0;
  }
}

export class AtlasProcessor {
  constructor(atlasContent) {
    this.atlasContent = atlasContent;
    this.pages = [];
    this.regions = {};       // name → AtlasRegion (ordered by insertion)
    this._loadedImages = {}; // pageName → HTMLImageElement
    this._pageMap = {};      // pageName → AtlasPage
    this._parse();
  }

  _parse() {
    const lines = this.atlasContent.split('\n').map(l => l.trim());
    let currentPage = null;
    let currentRegion = null;

    for (const line of lines) {
      if (!line) continue;

      if (line.endsWith('.png')) {
        currentPage = new AtlasPage(line);
        this.pages.push(currentPage);
        this._pageMap[line] = currentPage;
        currentRegion = null;
        continue;
      }

      if (line.includes(':')) {
        const idx = line.indexOf(':');
        const key = line.slice(0, idx).trim().toLowerCase();
        const vals = line.slice(idx + 1).split(',').map(v => v.trim());

        if (currentRegion) {
          if (key === 'bounds' && vals.length >= 4) {
            currentRegion.x = parseInt(vals[0]);
            currentRegion.y = parseInt(vals[1]);
            currentRegion.w = parseInt(vals[2]);
            currentRegion.h = parseInt(vals[3]);
          } else if (key === 'xy') {
            currentRegion.x = parseInt(vals[0]);
            currentRegion.y = parseInt(vals[1]);
          } else if (key === 'size' && currentRegion.w === 0) {
            // Only apply size to region if we haven't got bounds yet
            currentRegion.w = parseInt(vals[0]);
            currentRegion.h = parseInt(vals[1]);
          } else if (key === 'rotate') {
            const v = vals[0].toLowerCase();
            if (v === 'true') currentRegion.rotate = 90;
            else if (v === 'false') currentRegion.rotate = 0;
            else { const n = parseInt(v); currentRegion.rotate = isNaN(n) ? 0 : n; }
          } else if (key === 'offsets' && vals.length >= 4) {
            currentRegion.offsets = vals.map(Number);
          } else if (key === 'index') {
            currentRegion.index = parseInt(vals[0]);
          }
        } else if (currentPage) {
          if (key === 'size') {
            currentPage.width = parseInt(vals[0]);
            currentPage.height = parseInt(vals[1]);
          } else if (key === 'format') {
            currentPage.format = vals[0];
          } else if (key === 'filter') {
            currentPage.filter = [vals[0], vals[1]];
          } else if (key === 'repeat') {
            currentPage.repeat = vals[0];
          }
        }
        continue;
      }

      // No colon and not a .png line → region name
      if (!currentPage) continue;
      currentRegion = new AtlasRegion(line, currentPage.filename);
      this.regions[line] = currentRegion;
    }
  }

  /**
   * Load images from a map of { pageName: File | string(dataURL|url) }.
   * Must be called before extracting regions.
   */
  async loadImages(imageFileMap) {
    for (const [pageName, source] of Object.entries(imageFileMap)) {
      try {
        const img = await _loadImage(source);
        const page = this._pageMap[pageName];
        if (page && page.width !== 0 && page.height !== 0) {
          if (img.naturalWidth !== page.width || img.naturalHeight !== page.height) {
            page.scaleX = img.naturalWidth / page.width;
            page.scaleY = img.naturalHeight / page.height;
          }
        }
        this._loadedImages[pageName] = img;
      } catch (e) {
        console.error(`Failed to load image ${pageName}:`, e);
      }
    }
  }

  getPageImage(pageName) {
    if (pageName) return this._loadedImages[pageName] || null;
    const keys = Object.keys(this._loadedImages);
    return keys.length > 0 ? this._loadedImages[keys[0]] : null;
  }

  /**
   * Crop a region from img (HTMLImageElement or canvas) and undo atlas rotation.
   * Returns a canvas element (w × h) with the sprite in its original orientation.
   */
  static cropAndRotate(img, x, y, w, h, rotate) {
    const isSwapped = rotate === 90 || rotate === 270;
    const cropW = isSwapped ? h : w;
    const cropH = isSwapped ? w : h;

    const canvas = document.createElement('canvas');
    canvas.width = w;
    canvas.height = h;
    const ctx = canvas.getContext('2d');

    if (rotate === 0) {
      ctx.drawImage(img, x, y, cropW, cropH, 0, 0, w, h);
    } else if (rotate === 90) {
      // Stored in atlas as h×w; un-rotate 90° CW → w×h
      ctx.translate(w, 0);
      ctx.rotate(Math.PI / 2);
      ctx.drawImage(img, x, y, cropW, cropH, 0, 0, cropW, cropH);
    } else if (rotate === 270) {
      // Stored in atlas as h×w; un-rotate 90° CCW → w×h
      ctx.translate(0, h);
      ctx.rotate(-Math.PI / 2);
      ctx.drawImage(img, x, y, cropW, cropH, 0, 0, cropW, cropH);
    } else if (rotate === 180) {
      ctx.translate(w, h);
      ctx.rotate(Math.PI);
      ctx.drawImage(img, x, y, w, h, 0, 0, w, h);
    }

    return canvas;
  }

  /** Extract a single region as a canvas (includes offset padding). */
  extractRegion(name) {
    const region = this.regions[name];
    if (!region) return null;
    const baseImg = this._loadedImages[region.pageFilename];
    if (!baseImg) return null;

    let { x, y, w, h, rotate, offsets } = region;
    const page = this._pageMap[region.pageFilename];

    if (page && (page.scaleX !== 1.0 || page.scaleY !== 1.0)) {
      x = Math.round(x * page.scaleX);
      y = Math.round(y * page.scaleY);
      w = Math.round(w * page.scaleX);
      h = Math.round(h * page.scaleY);
    }

    const sprite = AtlasProcessor.cropAndRotate(baseImg, x, y, w, h, rotate);
    const currentW = sprite.width;
    const currentH = sprite.height;

    if (offsets) {
      let [offX, offY, origW, origH] = offsets;
      if (page && (page.scaleX !== 1.0 || page.scaleY !== 1.0)) {
        offX = Math.round(offX * page.scaleX);
        offY = Math.round(offY * page.scaleY);
        origW = Math.round(origW * page.scaleX);
        origH = Math.round(origH * page.scaleY);
      }
      const canvas = document.createElement('canvas');
      canvas.width = origW;
      canvas.height = origH;
      const ctx = canvas.getContext('2d');
      ctx.drawImage(sprite, offX, origH - offY - currentH);
      return canvas;
    }

    return sprite;
  }

  /** Extract a single region as a base64 PNG data URI. */
  extractRegionAsDataURL(name) {
    const canvas = this.extractRegion(name);
    return canvas ? canvas.toDataURL('image/png') : null;
  }

  /**
   * Get a composite preview of one or more region names as a data URI.
   * Regions are composited (alpha-blended) on a max-size canvas.
   */
  getPreviewDataURL(names) {
    const images = names
      .filter(n => n in this.regions)
      .map(n => this.extractRegion(n))
      .filter(Boolean);

    if (images.length === 0) return null;
    if (images.length === 1) return images[0].toDataURL('image/png');

    const maxW = Math.max(...images.map(c => c.width));
    const maxH = Math.max(...images.map(c => c.height));

    const canvas = document.createElement('canvas');
    canvas.width = maxW;
    canvas.height = maxH;
    const ctx = canvas.getContext('2d');

    // Composite in reverse order (last on top matches Python's alpha_composite)
    for (const img of [...images].reverse()) {
      ctx.drawImage(img, 0, 0);
    }

    return canvas.toDataURL('image/png');
  }

  /**
   * Extract all regions. Returns { name: canvas }.
   */
  extractAll() {
    const results = {};
    for (const name of Object.keys(this.regions)) {
      try {
        const canvas = this.extractRegion(name);
        if (canvas) results[name] = canvas;
      } catch (e) {
        console.error(`Failed to extract ${name}:`, e);
      }
    }
    return results;
  }
}

// ─── Helpers ────────────────────────────────────────────────────────────────

/** Load an image from a File object or a URL/data-URL string. */
export function _loadImage(source) {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = () => reject(new Error(`Failed to load image: ${source}`));
    if (source instanceof File) {
      img.src = URL.createObjectURL(source);
    } else {
      img.src = source;
    }
  });
}
